nsq/reader.py

```diff
@@ -313,27 +314,43 @@ def _on_message(self, conn, message, **kwargs):
     def _handle_message(self, conn, message):
         self._maybe_update_rdy(conn)
 
-        success = False
+        result = False
         try:
             if 0 < self.max_tries < message.attempts:
                 self.giving_up(message)
                 return message.finish()
             pre_processed_message = self.preprocess_message(message)
             if not self.validate_message(pre_processed_message):
                 return message.finish()
-            success = self.process_message(message)
+            result = self.process_message(message)
         except Exception:
             logger.exception('[%s:%s] uncaught exception while handling message %s body:%r',
                              conn.id, self.name, message.id, message.body)
             if not message.has_responded():
                 return message.requeue()
 
-        if not message.is_async() and not message.has_responded():
-            assert success is not None, 'ambiguous return value for synchronous mode'
-            if success:
+        if result not in (True, False, None):
+            # assume handler returned a Future or Coroutine
+            message.enable_async()
+            fut = tornado.gen.convert_yielded(result)
+            fut.add_done_callback(functools.partial(self._maybe_finish, message))
```

Comment:
This is a subtle addition, but I like it! You can now implicitly enable async by returning a future or coroutine.

Probably something we should document?

Reply:
yeah ... but the callback attached to the future will call `message.finish()` automatically, so this acts kinda like a synchronous handler function ... I'll see about updating the docs

Reply:
ahh, right, nevermind